tests/dragonfly/json_test.py

```diff
@@ -0,0 +1,95 @@
+
+    try:
+        connection.json().set(key, path, value)
+
+        result = connection.json().get(key, path)
+        return result
+    except redis.exceptions.ResponseError as re:
+        return str(re)
+    except Exception as e:
+        return None
+
+def get_str(connect, key):
+    try:
+        return connect.get(key)
+    except redis.exceptions.ResponseError as re:
+        return str(re)
+    except Exception as e:
+        return None
+
+def test_basic_json_get_set(client):
+    key_name="test-json-key"
+    result = get_set_json(connection=client, key=key_name, value=jane)
+    if result is None:
+        assert False, "failed to set JSON value"
+    else:
+        the_type = get_type(connection=client, key=key_name)
+        assert the_type == "ReJSON-RL"
+        assert len(result) == 1
+        assert result[0]['name'] == 'Jane'
+        assert result[0]['Age'] == 33
+
+def test_access_json_value_as_string(client):
+    key_name="test-json-key"
+    result = get_set_json(connection=client, key=key_name, value=jane)
+    if result is None:
+        assert False, "failed to set JSON value"
+    else:
+        # make sure that we have valid JSON here
+        the_type = get_type(connection=client, key=key_name)
+        assert the_type == "ReJSON-RL"
+        # you cannot access this key as string
+        result = get_str(connect=client, key=key_name)
+        assert result is not None
+        assert result == "WRONGTYPE Operation against a key holding the wrong kind of value"
+
+def test_reset_key_to_string(client):
+    key_name="test-json-key"
+    result = get_set_json(connection=client, key=key_name, value=jane)
+    if result is None:
+        assert False, "failed to set JSON value"
+    else:
+        # make sure that we have valid JSON here
+        the_type = get_type(connection=client, key=key_name)
+        assert the_type == "ReJSON-RL"
+        # set the key to be string - this is legal
+        assert client.set(key_name, "some random value"), "we should be able to set type to string"
+        result = get_str(connect=client, key=key_name)
+        assert result == "some random value"
+        # For JSON set the update the root path, we are allowing
+        # to change the type to JSON and override it
+        result = get_set_json(connection=client, key=key_name, value=jane)
+        the_type = get_type(connection=client, key=key_name)
+        assert the_type == "ReJSON-RL"
+
+def test_update_value(client):
+    key_name="test-json-key"
+    result = get_set_json(connection=client, key=key_name, value=json_num)
+    if result is None:
+        assert False, "failed to set JSON value"
+    else:
+        # make sure that we have valid JSON here
+        the_type = get_type(connection=client, key=key_name)
+        assert the_type == "ReJSON-RL"
+        result = get_set_json(connection=client, value="0", key=key_name, path="$.a.*")
+        assert len(result) == 3
```

Comment:
Whats the main difference between this and the json unit tests?

The tests are fine, I just imagined that we would use pytests for smth more advanced to test overall integrity.

Reply:
I think that there is 2 reasons:
1. We're using some external tool to verify that the support is correct, When using these tools you sometimes find issue with return types or the args passing, in the unit tests you're doing a little overfitting in the sense that we are didn't verify that we're returning or getting what external tools are sending to us (for example, the example in redis doc for JSON would not work with us, since it sends invalid root as a path..).
2. I think that unit tests should focus on the unit itself - the effect the json commands only, U tried to make sure that with other other types we're working correctly, and less and spesific command.

Reply:
I think that most logic should be inside unit tests. It's fine to do sanity checks and see we are compatible with the extisting clients (aka py-redis) but I would not duplicate logic testing into pytests.

I find them useful to provide coverage for input/output (sockets/Resp protocol) that is not covered by unit tests

Reply:
I tried to make the python tests short for this reason, I didn't want to duplicate unit tests, and I didn't do many command validations in this case, since this is covered with the unit tests. I focus on what I found when I ran some commands from the redis-cli to verify that we are matching Redis. The only real use case for python here is to verify that we can set JSON and we cannot read or it as string and that when we are updating something that was change to string we are getting an error. I guess we can do the same with unit test, but the main reason in this case is to verify that we are working correctly with external tool (for example reporting the correct type to it).